### PR TITLE
RichText: selectionChange: bind on focus, unbind on blur

### PR DESCRIPTION
**Pull request by ellatrix**
## Description

This PR is an attempt to decrease the work done on a key press. Every key press is eventually followed by (several) selection change events. Every `RichText` instance has a listener on this global event, all of which will be called multiple times after a key press. This is not significant for a few instances, but it becomes quite noticeable if there are a huge amount of instances.

The solution is to only add the listener on focus, then remove again on blur.

Testing on with [this post](https://github.com/WordPress/gutenberg/issues/10418#issuecomment-436431030) I notice one selection change event takes on average 9ms to complete. With this PR it takes on average 0.3ms.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] My code has proper inline documentation.

### CODE REVIEW

**Comment by iseulde** [comment]
screen shot 2018-11-30 at 20 48 30

In this graph you can see all the little callbacks on every selection change.

File: packages/editor/src/components/rich-text/index.js [hunk 1]
```diff
@@ -354,6 +347,12 @@ export class RichText extends Component {
 		if ( unstableOnFocus ) {
 			unstableOnFocus();
 		}
+
+		document.addEventListener( 'selectionchange', this.onSelectionChange );
+	}
+
+	onBlur() {
+		document.removeEventListener( 'selectionchange', this.onSelectionChange );
```
Comment (aduth) [inline packages/editor/src/components/rich-text/index.js]: I guess `blur`, despite bubbling, doesn't fire when moving between elements in the same `contenteditable` ?

https://codepen.io/aduth/pen/VVqOmq?editors=1111

(In other words, I suspected there could be an issue, but it doesn't appear so)

Comment (iseulde) [reply packages/editor/src/components/rich-text/index.js]: No, it shouldn't, it's starts at the `contenteditable` element.

File: packages/editor/src/components/rich-text/index.js [hunk 2]
```diff
@@ -354,6 +347,12 @@ export class RichText extends Component {
 		if ( unstableOnFocus ) {
 			unstableOnFocus();
 		}
+
+		document.addEventListener( 'selectionchange', this.onSelectionChange );
```
[the remark this answers is not in the document]
Comment (iseulde) [reply packages/editor/src/components/rich-text/index.js]: Yeah, we should be able to remove this.

File: packages/editor/src/components/rich-text/index.js [hunk 3]
```diff
@@ -392,11 +386,6 @@ export class RichText extends Component {
 	 * Handles the `selectionchange` event: sync the selection to local state.
 	 */
 	onSelectionChange() {
-		// Ensure it's the active element. This is a global event.
-		if ( ! this.isActive() ) {
```
Comment (aduth) [inline packages/editor/src/components/rich-text/index.js]: I think I'd noted it elsewhere, but there's a small difference between `onFocus` and `isActive` in that the prior was testing the `div` wrapper as the active element, the latter exclusively applied to the TinyMCE element (i.e. not considering toolbars and such). When last I tested this, I recall it being a reasonable difference in that (a) all other elements are non-visual or slot-rendered and (b) are not relevant to the selection anyways.

Does that sound accurate to you?

Comment (iseulde) [reply packages/editor/src/components/rich-text/index.js]: Hm, I think both are checking the `contenteditable` element? `this.editableRef` is the `contenteditable` element, and `onFocus` is a listener on that element. Maybe you're thinking about `setFocusedElement`?

Comment (aduth) [reply packages/editor/src/components/rich-text/index.js]: > Hm, I think both are checking the `contenteditable` element? `this.editableRef` is the `contenteditable` element, and `onFocus` is a listener on that element. Maybe you're thinking about `setFocusedElement`?

Ah! I am!

I still think we should consolidate these, as having multiple concepts of focus here is particularly confusing for me, but it's not relevant to the change proposed here.

Comment (iseulde) [reply packages/editor/src/components/rich-text/index.js]: Yeah, tbh I'm not sure why there's something like `setFocusedElement` on the wrapper...

**Review by aduth — APPROVED** [review]
👍

File: packages/editor/src/components/rich-text/index.js (hunk 3 above)
Comment (aduth) [reply packages/editor/src/components/rich-text/index.js]: My hope was that at some point, we could get rid of the wrapper, at least in the sense of having it be a `<Fragment>` instead of an actual DOM `div` node.

**Comment by gziolo** [comment]
@iseulde, out of curiosity, do you have an idea if there would be a way to detect that you are still editing a given `RichText` when you use block's toolbar? Use case would be as follows:
1. I type something in the `RichText` area (focus is on RichText)
1. I use the keyboard shortcut to start navigating in the toolbar (blur is triggered on RichText)
1. I pick the formatting option
1. I get back to editing in `RichText` (focus gets back to RichText)
1. I select another text in `RichText
1. I use the keyboard shortcut to add link (focus would stay in RichText because popover is in the same virtual dom tree)
1. I save the link
1. I get back to editing in `RichText`

In other words, whatever you do related to editing, we would need to be able to identify a currently active `RichText` field. As far as I was trying to tackle it, I had the issue with the focus/blur dance when using block's toolbar. I think the same issue would apply in case you want to use `InsepectorControls` like in your plugin which adds colors formatting. Inline tokens like images, might be another use case where this is getting tricky.

**Comment by iseulde** [comment]
Sounds like this is what `setFocusedElement` in the discussion above?

**Comment by gziolo** [comment]
> Sounds like this is what `setFocusedElement` in the discussion above?

This is where we want it to be. It's not yet there :( The issue is mostly when there are other components like `PlainText` which should take over focus and reset formatters from `RichText`. It isn't the case. We have an issue which tracks it.

**Comment by iseulde** [comment]
But this PR doesn't cause any issues right?

**Comment by gziolo** [comment]
> But this PR doesn't cause any issues right?

I don't think so. I just wanted to discuss further improvements :)

**Comment by aduth** [comment]
There's an issue with the changes here, in that if the component unmounts while it's focused and before it's blurred, an error can be logged. This can be seen when merging two blocks, for example.

1. Navigate to Posts > Add New
2. Add a paragraph with some text
3. Add a second paragraph with some text
4. Place caret at the beginning of second paragraph and press Backspace

I'll plan to submit a fix soon.

**Comment by aduth** [comment]
> I'll plan to submit a fix soon

See: #12817 (now merged)
